app/src/main/java/co/tinode/tindroid/db/MessageDb.java

```diff
@@ -261,93 +308,189 @@ static Cursor queryUnsent(SQLiteDatabase db, long topicId) {
-                                               int fromId, int toId, List<Integer> list, boolean markAsHard) {
-        int affected = 0;
+    private static boolean deleteOrMarkDeleted(SQLiteDatabase db, long topicId, int delId, int fromId, int toId,
+                                               boolean markAsHard) {
+        // 1. Delete all messages within the give range (sent and unsent).
+        // 2. Delete all unsynchronized (soft and hard) deletion ranges fully within this range
+        // (no point in synchronizing them, they are superseded).
+        // 3.1 If server record, consume older partially overlapping server records,
+        // 3.2 If client hard-record, consume partially overlapping client hard-records.
+        // 3.3 If client soft-record, consume partially overlapping client soft records.
+        // 4. Expand current record to consumed range.
+
+        boolean success = false;
+
+        // Message selector: all messages in a given topic with seq between fromId and toId [inclusive, exclusive).
+        String messageSelector = COLUMN_NAME_TOPIC_ID + "=" + topicId;
+        ArrayList<String> parts = new ArrayList<>();
+        if (fromId > 0) {
+                parts.add(COLUMN_NAME_SEQ + ">=" + fromId);
+        }
+        parts.add(COLUMN_NAME_SEQ + "<" + toId);
+        if (parts.size() > 0) {
+            messageSelector +=  " AND " + TextUtils.join(" AND ", parts);
+        }
+        messageSelector += " AND " + COLUMN_NAME_STATUS + "<=" + BaseDb.Status.SYNCED.value;
+
+        // Selector of ranges which are fully within the new range.
+        parts.clear();
+        String rangeDeleteSelector = COLUMN_NAME_TOPIC_ID + "=" + topicId;
+        if (fromId > 0) {
+            parts.add(COLUMN_NAME_SEQ + ">=" + fromId);
+        }
+        parts.add(COLUMN_NAME_HIGH + "<" + toId);
+        if (parts.size() > 0) {
+            rangeDeleteSelector +=  " AND " + TextUtils.join(" AND ", parts);
+        }
+        // All types: server, soft and hard.
+        rangeDeleteSelector += " AND " + COLUMN_NAME_STATUS + ">" + BaseDb.Status.SYNCED.value;
+
+
+        // Selector of partially overlapping deletion ranges. Find bounds of existing deletion ranges of the same type
+        // which partially overlap with the new deletion range.
+        String rangeConsumeSelector = COLUMN_NAME_TOPIC_ID + "=" + topicId;
+        BaseDb.Status status;
+        if (delId > 0) {
+            rangeConsumeSelector += " AND " + COLUMN_NAME_DEL_ID + "<" + delId;
+            status = BaseDb.Status.DELETED_SYNCED;
+        } else {
+            status = markAsHard ? BaseDb.Status.DELETED_HARD : BaseDb.Status.DELETED_SOFT;
+        }
+        rangeConsumeSelector += " AND " + COLUMN_NAME_STATUS + "=" + status.value;
+
+        String rangeNarrow = "";
+        parts.clear();
+        if (fromId > 0) {
+            parts.add(COLUMN_NAME_HIGH + ">=" + fromId);
+        }
+        parts.add(COLUMN_NAME_SEQ + "<" + toId);
```

Comment:
Should it be non-strict here?
Here's a case: if `COLUMN_NAME_SEQ` is [4,..) and `toId` == 4, you can treat the former range as overlapping (adjacent) with the one where toId = 4.

Reply:
Done